text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+- A similar ambiguity exists between an "object" (a data structure denoted by key/value pairs) and a table of one row
+- Inner-tables are modelled differently than top-level tables, leading to confusion
+- There is no way to currently represent a matrix
+- As rows are streamed instead of tables, it's not possible to predict how to display this information. This is currently mitigated by buffering some number of rows and treating them as one "table"
+- Likewise, since rows are streamed instead of tables, it's unclear what the user should expect if they request a column that is not present, as this column may appear in the following row instead.
+- When table data is serialized, there is a large amount of duplication, as columns are repeated with each row sent.
+- Additionally, there is currently no way to represent rows using a row literal. We propose a frame literal that will fill this role.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Data frame representation:
+
+```rust
+struct DataFrame {
+    headers: Option<Vec<String>>,
+    rows: Vec<Vec<Value>>,
+    partial_frame_id: Option<Uuid>,
+    is_object: bool,
+}
+```
+
+## Tables
+
+A self-contained table would look like this:
+
+```rust
+let frame = DataFrame {
+    headers: Some(vec!["name", "age"]),
+    rows: vec![
+        vec![Value::from("Bob"), Value::from(30)],
+        vec![Value::from("Sally"), Value::from(43)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+};
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: [Bob, Sally], age: [30, 43]]
+```
+
+## Lists and matrices
+
+A self-contained list would look like this:
+
+```rust
+let list = DataFrame {
+    headers: None,
+    rows: vec![
+        vec![Value::from(1), Value::from(2), Value::from(3)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+}
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[1, 2, 3]
+```
+
+## Objects (aka hash tables)
+
+A self-contained object would look like this:
+
+```rust
+let obj = DataFrame {
+    headers: Some(vec!["name", "level"]),
+    rows: vec![
+        vec![Value::from("Thomas"), Value::from(12)]
+    ],
+    partial_frame_id: None,
+    is_object: true,
+}
+```
+
+**Note:** we use the boolean in the table rather than enumeration because all processing on the frame remains uniform regardless of if the frame is a single row with headers vs an object. This simplifies algorithms to only have to work with the data directly, and we can later represent this data and/or serialize this data in a way that maintains the user's model.
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: Thomas, level: 12]
+```
+
+## Streaming
+
+An important part of Nu is being able to work with potentially endless streams of data. As is often the case when working with external commands, there's no guarantee that the output will terminate.
```

Comment:
If we're serious about supporting the data science community, I don't think we can only stream data. In real world scenarios, your data is nearly always flawed in some way. We'll need to be able to check the data with something like pandas pd.info(), be able to remove dupes, be able to remove columns, remove nulls or incomplete rows, and generally manipulate the data before passing it on to the next pipeline, or even saving the now purified data. I think this is crucial for data science. Perhaps this could be accomplished via some `--no-stream` parameter that allows us to consume the entire data set and manipulate it via pipelines. Or maybe this is a phase 2 thing after the more generalized dataframe is in place.

Reply:
You aren't required to stream the dataframes. One way this design differs from the current one is that you can send a whole table at once, as a single data frame.

I wanted to make sure it was possible to stream the frames, but this isn't required.